lib/coffee_place/importer.rb

```diff
@@ -0,0 +1,87 @@
+# frozen_string_literal: true
+
+require 'csv'
+require 'open-uri'
+
+require_relative './result'
+require_relative './location'
+
+module CoffeePlace
+  # Imports locations from local or remote CSV file.
+  class Importer
+    def initialize
+      @locations = []
+      @errors = []
+    end
+
+    def import_from(source_name)
+      line_index = 0
+
+      with_source(source_name) do |csv|
+        CSV.parse(csv) do |line|
+          line_index += 1
+          import_line(line, line_index)
+        end
+      end
+
+      location_results
+    end
+
+    def import_line(line, line_index)
+      place_name, lat, lon = line
+
+      result = Location.validate(name: place_name, lat: lat, lon: lon)
+
+      if result.success?
+        @locations << result.value
+      else
+        add_import_error "CSV error on line #{line_index}: #{result.error}"
+      end
+    end
+
+    def location_results
+      if @errors.empty?
+        Result.success(@locations)
+      else
+        Result.failure(@errors)
+      end
+    end
+
+    def with_source(source_name, &block)
+      uri = URI.parse(source_name)
+
+      # When we have an URL, download remote file
+      return with_remote_file(uri, &block) if uri.scheme
+
+      # When we have a local file, try to open it
+      return with_local_file(uri, &block) if File.exist?(uri.path)
+
+      add_import_error unknown_source_error_message(source_name)
+    rescue URI::InvalidURIError
+      add_import_error "#{source_name} is not a valid URL or local filename"
+    end
+
+    def with_remote_file(uri, &block)
+      uri.open(&block)
+    rescue OpenURI::HTTPError => e
+      add_import_error "Failed to download remote file \"#{uri}\": #{e.message}"
```

Comment:
What if an invalid URI is provided, e.g. `http://asd.asd`?

Reply:
The method from where `with_remote_file` is called - `with_source` has a rescue as well.

```ruby
    def with_source(source_name, &block)
      uri = URI.parse(source_name)

      # When we have an URL, download remote file
      return with_remote_file(uri, &block) if uri.scheme

      # When we have a local file, try to open it
      return with_local_file(uri, &block) if File.exist?(uri.path)

      add_import_error unknown_source_error_message(source_name)
    rescue URI::InvalidURIError
      add_import_error "#{source_name} is not a valid URL or local filename"
    end
```

That `URI::InvalidURIError` catches invalid URL's.
What I try to do here is to see if it is a parsable (i.e. valid) URI.
If it is an URI, then see if it has a `scheme` - as all `http` and `https` URLs do.
If it is valid as an URI, but not also a URL, then blindly assume we are dealing with a file,
and try to load the CSV from that local file.

Ways this could horribly backfire: someone sends some URL with a different scheme,
e.g. a database scheme: `redis://<some_ip>/whoops`

This could potentially become a security vulnerability.
More sanity checks should be done on the input, before trying to open that uri.

In a real application, of course, I would place a todo at the top of the method:
```ruby
  # TODO: make sure we aren't hacked in a stupid way  (commented by madalindaniel 11months ago) ... 
```

:stuck_out_tongue:

Reply:
Basically, whenever you download some remote file to your server ... be weary!
The night is dark and full of errors. :ghost:

Reply:
Sorry, I didn't read the example you provided carefully the first time.
`http://asd.asd` is not an invalid URI. 
`asd` is simply not yet defined as a top level domain.

I think `uri.open` will result in a `DNS` query trying to find the IP responsible for this domain, and fail,
resulting in a  `OpenURI::HTTPError` error being raised.

Reply:
Whelp, just tested it ... it throws a wonderful:
3.0.0/net/http.rb:987:in `initialize': getaddrinfo: Name or service not known (SocketError)

I am not catching that error, and it is bubbling through both rescue clauses.
I was hoping Ruby's net/http catches and re-throws these as `OpenURI::HTTPError`, but it doesn't.
I'll add a few more rescue clauses to catch `SocketError` as well ... sigh ...

Reply:
Update: fixed by [89e854](https://github.com/Agilefreaks/test_oop/pull/23/commits/89e854c4b9a5371b886788878bdab7488707d745)

Thank you for showing me an edge case - nice save!

![nice_save](https://user-images.githubusercontent.com/7295751/132101181-8e4ea3ff-9012-4e06-a89d-3b9549a132f5.png)

Reply:
Haha, no problem :D